### PR TITLE
Packit/TMT: upstream maintenance of downstream gating tests

### PR DESCRIPTION
**Pull request by lsm5**
This commit reorganizes tests such that tests can be easily triggered on local environment without any specialized CI setup as well as triggered by TMT on PRs and bodhi updates. This essentially moves all Fedora and CentOS Stream downstream testing maintenance to the upstream repo.

Tests on upstream PRs use packages from the `podman-next` copr wherever relevant while downstream tests use packages from the official distro repos.
    
This commit also separates out rhel and centos stream tasks.

This enables centos-stream jobs to run automatically on all PRs regardless of the PR author's access level on the repo. RHEL tests on PRs by non-maintainers would need manual triggereing by maintainers.

### CODE REVIEW

File: plans/all.fmf
Comment (LecrisUT) [inline plans/all.fmf:22]: This can be moved to `test` section. You should test it though, some `context` are not available at that stage, but I believe `initiator` will always be available.

File: plans/all.fmf
Comment (LecrisUT) [inline plans/all.fmf:14]: How about using `tag` filter instead of name ones? than you can just pick up specific tests from tags on upstream and donwstream

Comment (LecrisUT) [inline plans/all.fmf:14]: About these tests, can they be run sequentially one after another? If so it would make more sense to use a single plan with multiple tests, will also display nicer in `testing-farm`

File: rpm/gating.yaml
Comment (LecrisUT) [inline rpm/gating.yaml:3]: The `gating.yaml` I never found a good documentation on it, let me know if you have any.

File: .packit.yaml
Comment (LecrisUT) [inline .packit.yaml:16]: `files_to_sync` change is coming later I guess?

File: plans/all.fmf
Comment (LecrisUT) [inline plans/all.fmf:6]: nitpick, I usually don't put `+` on the lowest level, just to make sure I am not inheriting from something I don't expect, e.g. forgetting a `main.fmf` in the source.

File: test/main.fmf
Comment (LecrisUT) [inline test/main.fmf:9]:
```suggestion
require:
    - make
    - policycoreutils

test: "make "

/basic_check:
    summary: Run basic checks
    test+: basic_check
```

You can do stuff like these. Not sure if it makes sense here or not. Another design I had was to use `test: make $TMT_TEST_NAME`, but that might need some refactoring. Personally though I tend to stay away from Makefile, but that's a personal taste.

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:34]: Can move this to `require`?

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:31]: Can add more comments on why this one is needed?

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:26]: These you can add in the `plan` section
```yaml
prepare:
    how: install
    copr: rhcontainerbot/podman-next
    package: podman
```

Adding multiple copr, I am not sure, but you can try to add a similar `prepare` step with empty `package`.

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:42]: Does it affect if there are extra packages installed on the other tests, e.g. `golang` on `system`? This might change how these could be designed

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:51]:
```suggestion
    if rpm -q dnf5; then
```
What about making it more compact? I guess you can go even more so if you don't care about readability
```bash
rpm -q dnf5 && SRPM_OPTS="--srpm" || SRPM_OPTS="--source"
```

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:67]: Oh lol, maybe a future challenge, making these into `tmt` tests on `podman` and then simplifying it to `discover.url: https://github.com/containers/podman`. But one step at a time :D

I haven't investigated go design, but python is pretty straightforward to design an compartmentalize these.

File: test/podman-tests.sh
Comment (LecrisUT) [inline test/podman-tests.sh:76]: This can be controlled with [`feature`](https://tmt.readthedocs.io/en/stable/spec/plans.html#feature), but I think it's by default on `testing-farm`, not sure about RHEL though. You can alter the order of `prepare` to make sure that only `bats` is downloaded from epel.

**Review by LecrisUT — COMMENTED** [review]
Overall review, you can move much of the shell script to the plans, particularly the installation sections, and then synchronize them with `files_to_sync`. It is a bit of a chicken-and-egg issue with the `dist-git-source`, but what you can do to test these is to create a patch file in the dist-git repo and add the changes of this PR to it

File: test/podman-tests.sh
Comment (lsm5) [reply test/podman-tests.sh:26]: That doesn't work if i want to enable centos-stream copr targets on centos-stream envs. All EL envs end up choosing epel by default.

Sure, we could do multiple prepare steps, but then we end up with too many prepare steps and there's concern that we may miss some distro-version-arch combo. It would probably be much easier if a single `adjust` could handle if-else in addition to a `default & exception` way.

File: test/podman-tests.sh
Comment (lsm5) [reply test/podman-tests.sh:31]: because it's a PITA? :)  . I've found the testing-farm repos to always interfere with the packages I want to use. Things are a lot easier if I just delete it.

File: test/podman-tests.sh
Comment (lsm5) [reply test/podman-tests.sh:51]: yes, even better. I'll add that. thanks

File: test/podman-tests.sh
Comment (lsm5) [reply test/podman-tests.sh:76]: i've had tons of trouble making epel work on internal testing farm. So, I was looking for something that'd just work across all testing farms. IIRC, bats rpm pulls in parallel which also ends up pulling in a bunch of perl packages. So, I wonder if `only bats` would work.

I would've preferred to just install bats from upstream using their install script, but some people in the team had a preference for rpms.

File: .packit.yaml
Comment (lsm5) [reply .packit.yaml:16]: yes

File: plans/all.fmf
Comment (lsm5) [reply plans/all.fmf:14]: RE: sequentially, the system and e2e tests can run after the basic_check, but e2e and system tests I'd probably not want them run sequentially even if that's doable. For now, testing-farm does show 3 sections `basic, e2e and system`  though, which I guess is good enough

File: test/main.fmf
Comment (lsm5) [reply test/main.fmf:9]: main reason to use Makefile was for a human to quickly run some make command, without having to bother with TMT. For that same reason, I guess it's best to have the package dependencies / copr enablement conditionals within the script itself along with some warnings whenever a copr or epel is being enabled.

But I do hear your suggestions, thanks for those. I'll probably switch to idiomatic packit if everyone is comfortable enough with TMT.

**Comment by lsm5** [comment]
> Overall review, you can move much of the shell script to the plans, particularly the installation sections, and then synchronize them with `files_to_sync`. It is a bit of a chicken-and-egg issue with the `dist-git-source`, but what you can do to test these is to create a patch file in the dist-git repo and add the changes of this PR to it

@LecrisUT really appreciate all the review comments. The main reasons to not use idiomatic TMT for installation and setup is so that people could run tests with a simple `make ...` or `./test/podman-tests.sh ..` without worrying about TMT setup. Majority of the podman team isn't familiar with TMT.  And one of the team's requirements for our ongoing CI refactor is to run tests on local env without any specialized setup.

File: plans/all.fmf
Comment (lsm5) [reply plans/all.fmf:14]: I added the `upstream and downstream` tag filters in the latest version. Seems to work and simplifies all.fmf.

It's hard for me to tell, but is it the case that all tests with a particular tag get run sequentially? Or are they still parallelized wherever possible?

**Comment by lsm5** [comment]
@LecrisUT looks like zuul / fedora-ci doesn't support tags yet?? Zuul can't find anything at: https://src.fedoraproject.org/rpms/container-selinux/pull-request/87

**Comment by LecrisUT** [comment]
> @LecrisUT looks like zuul / fedora-ci doesn't support tags yet?? Zuul can't find anything at: https://src.fedoraproject.org/rpms/container-selinux/pull-request/87

It should work fine, [here](https://src.fedoraproject.org/rpms/python-scikit-build-core/blob/rawhide/f/plans/examples.fmf)'s a reference. Getting the snytax of the filter can be tricky, but I see tmt upstream [uses the same](https://github.com/search?q=repo%3Ateemtee%2Ftmt+filter%3A+tag&type=code) as you, so I don't think that's it.

I think the issue is with `dist-git-source: true` alone does not extract the archive correctly :thinking:. Try to add [`dist-git-extract: container-selinux-*/`](https://src.fedoraproject.org/rpms/python-scikit-build-core/blob/rawhide/f/plans/main.fmf) (dunno what format the archive wil have there).

> @LecrisUT really appreciate all the review comments. The main reason to not use idiomatic TMT for installation and setup is so that people could run tests with a simple make ... or ./test/podman-tests.sh .. without worrying about TMT setup. Majority of the podman team isn't familiar with TMT. And one of the team's requirements for our ongoing CI refactor is to run tests on local env without any specialized setup.

Understandable. I believe one goal of tmt is to make running the tests consistent and easy on all environments, but of course there's still some work to be done on the UX and it is still a learning curve. There are some nice simplifications in a full tmt design (in this case w.r.t. installation and running on any OS environment), but it is understandable that this would be too much of a steep transition atm. I will focus on keeping it locally reproducible.

File: plans/all.fmf
Comment (LecrisUT) [reply plans/all.fmf]: > It's hard for me to tell, but is it the case that all tests with a particular tag get run sequentially? Or are they still parallelized wherever possible?

tmt still does not have the concept of parallelizing test runs. Can assume it will be sequentially for the foreseeable future.

Comment (LecrisUT) [reply plans/all.fmf]: Each test has it's own work directory, but they share the same machine, and the same environment produced by `plan.prepare`. That is mostly what I was referring on there. Do you want them to be completely independent, in which case, the original approach makes more sense, or do you want to reuse the `install` sections a bit to save up on some plan maintenance (having only upstream/downstream plans instead)

File: test/main.fmf
Comment (LecrisUT) [reply test/main.fmf:9]: Cool, however the team feels more comfortable. The trade here is mostly having to maintain the installation and logging manually, vs delegating to tmt.

File: test/podman-tests.sh
Comment (LecrisUT) [reply test/podman-tests.sh:36]: Lol, I guess these projects have quite some overlap with what testing-farm is doing for itself. But can ping testing-farm team to check these? At the point of when testing-farm is actually executing, the environment should be as vanilla as possible, and I think these changes should also be included in testing-farm by default.

File: test/podman-tests.sh
Comment (LecrisUT) [reply test/podman-tests.sh:31]: > It would probably be much easier if a single adjust could handle if-else in addition to a default & exception way.

Hmm, good point. I don't quite have a clean solution for this. The only thing I can think of is to have `fedora` (or a blank common) as a default (`else` variant) and use `prepare-:` to delete steps, and `prepare.[].order` to get a specific ordering.

But of course there is also the re-usability of the scripts, so it might not be relevant rn anyway.

File: test/podman-tests.sh
Comment (LecrisUT) [reply test/podman-tests.sh:76]: I mean not `only bats`, but rather pulling in at the appropriate time. But this was in the context of having multiple `prepare` steps. For the context of reproducibility, this would be preferred indeed.

File: test/main.fmf
Comment (LecrisUT) [inline test/main.fmf:10]:
```suggestion
    tag: [ downstream, upstream ]
```

This also works ;)

File: test/main.fmf
Comment (LecrisUT) [inline test/main.fmf:15]:
```suggestion
/upstream:
    tag: upstream
/upstream/podman_e2e_test:
    summary: Run SELinux specific Podman e2e tests on upstream PRs
    test: make podman_e2e_test_upstream
```

Other things you can do is to use the [`inheritance`](https://fmf.readthedocs.io/en/stable/features.html#inheritance) and [`elasticity`](https://fmf.readthedocs.io/en/stable/features.html#elasticity). Up to you how you think it will be easier to organize.

File: .packit.yaml
Comment (LecrisUT) [inline .packit.yaml:16]: These are `rsync` commands so it should be:
```suggestion
  - src: plans/
    dest: plans/
    delete: true
  - src: .fmf/
    dest: .fmf/
  - .packit.yaml
```
Also don't forget `.packit.yaml` as well, unless you want it seperately defined?

One thing that I've been bitten by is needing to delete a plan downstream, so make sure you have `delete: true` where needed.

**Review by LecrisUT — COMMENTED** [review]
This round LGTM, just some touchups on `tag` and probably needing `dist-git-extract`

File: test/main.fmf
Comment (lsm5) [reply test/main.fmf:10]: lol yes, when I tried that none of the tests showed up downstream, hence this experiment :D

File: plans/all.fmf
Comment (LecrisUT) [inline plans/all.fmf:18]:
```suggestion
    adjust+:
      - enabled: false
        when: initiator is not defined or initiator != packit
```
Oh this is a fun `fmf` aspect. When you had it here, `adjust` was overwriting one-another, so the `dist-git-source` one, had no effect`. Seems like you need to use the `adjust+:` here and make both of the `adjust` parts into lists

Comment (LecrisUT) [reply plans/all.fmf]: Still need converting the other `adjust` parts ;)

Comment (LecrisUT) [reply plans/all.fmf:12]: Oh lol, I guess you've found another solution. The other one would have been yo make the yaml value of `adjust` into a list. Then it's like python's `list += list` operation

**Comment by lsm5** [comment]
@LecrisUT I realized I could just sync `test/` over to downstream using files_to_sync and not bother with `dist-git-*` . We don't need to use Source0. Don't know if that'll hold true for all other projects related to podman but those could be handled later.

**Comment by LecrisUT** [comment]
Indeed, that's a good approach. As long as the tests are not too complicated or coupled with the Source0 data, it should work perfectly fine. I will ping on FedoraCI to see if the `dist-git-*` bug can be investigated further.

**Comment by lsm5** [comment]
@rhatdan @edsantiago @Luap99 @mheon PTAL.

So the current state of the PR works for gating tests:
- upstream on this PR (centos-stream is blocked on successful podman builds in podman-next copr and some other unrelated environment issues for those targets, can be ignored for now)
- [downstream on Zuul](https://src.fedoraproject.org/rpms/container-selinux/pull-request/87) with every dist-git PR
- [downstream on Bodhi updates](https://bodhi.fedoraproject.org/updates/FEDORA-2024-d3ac9f6271)

The Zuul and Bodhi test runs were verified this time by manually copying over files to dist-git. These will be automated via `.packit.yaml` (see below).

To make reviews hopefully easier:
- The podman revdep tests can be triggered on local envs without any TMT setup using test/Makefile or `test/podman-tests.sh`. So, please at least review those 2 files. Tests can be triggered using `make -C test podman_$TYPE_test_$STREAM` for types `e2e and system` and streams `upstream` and `downstream`.

- `plans/all.fmf` handles discovery and filtering of tests based on env (upstream / downstream) and initiator (human, packit, bodhi etc.), while test/main.fmf lists all tests along with tags that can be used for filtering.

- `.packit.yaml` now includes a `files_to_sync` which will copy over the listed dirs and files on every downstream package update and delete any stale files downstream.

- `gating.yaml` usually listed in Fedora/CentOS dist-git has been unified upstream in `rpm/gating.yaml`.

I plan to replicate this on our other repos once we have this approved and merged.

File: test/podman-tests.sh
Comment (edsantiago) [inline test/podman-tests.sh:4]: Comment says "do not set -e" immediately before "set -e"

File: test/podman-tests.sh
Comment (edsantiago) [reply test/podman-tests.sh:36]: I have to agree with @LecrisUT: this requires an explanation. Comments must explain _why_, not what, and they belong in the code, not in a github discussion.

File: test/podman-tests.sh
Comment (edsantiago) [inline test/podman-tests.sh:63]: EEEK. First: where are you cd'ed? This should be done from a scratch working dir, otherwise you don't know if you have space, or if you're vandalizing someone else's workspace, or if there are already existing `.tar.gz` files, or or or

File: test/podman-tests.sh
Comment (edsantiago) [inline test/podman-tests.sh:67]: This too is squicky: separate dnf installs of podman and podman-srpm; this means they could mismatch, and tests are only valid if they are 100% matched with the executable. I don't know how much of a real-world problem this will be or how to resolve it.

File: test/podman-tests.sh
Comment (edsantiago) [inline test/podman-tests.sh:11]: The conventional way to test this is `if [[ -z "$1" ]].....`

Also, please define named variables for these: 
```bash
TEST_TYPE=$1
STREAM=$2
```
Then all the code below can work using those names instead of the unclear `$1/$2`

File: test/podman-tests.sh
Comment (edsantiago) [inline test/podman-tests.sh:39]: If you need to do this in here (vs `require`) (and I don't know enough to comment on that), please install `podman-tests`, not podman. That will guarantee that tests match. And you can then remove the separate `dnf install podman-tests` below. You may need to do the `bats` setup/install earlier.

**Review by edsantiago — CHANGES_REQUESTED** [review]
A few concerns; and a lot of yaml code that I'm not qualified to review. Overall LGTM but needs a little cleanup. Thank you!

File: test/podman-tests.sh
Comment (lsm5) [reply test/podman-tests.sh:70]: `dnf download` will always fetch the latest podman srpm available which should match the latest available podman rpm as both cases will fetch from the same dnf repo (both fetch from podman-next copr or both fetch from official fedora).

If you prefer, I could add another check to compare versions of these and exit with error if mismatch.

**Review by edsantiago — APPROVED** [review]
LGTM, thank you!

**Comment by lsm5** [comment]
> LGTM, thank you!

@edsantiago thanks! @rhatdan PTAL.

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**
